// @nx/vite/src/generators/vitest/vitest-generator.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    default: function() {
        return _default;
    },
    vitestGenerator: function() {
        return vitestGenerator;
    },
    vitestGeneratorInternal: function() {
        return vitestGeneratorInternal;
    }
});
const _extends = require("@swc/helpers/_/_extends");
const _devkit = require("@nx/devkit");
const _js = require("@nx/js");
const _tssolutionsetup = require("@nx/js/src/utils/typescript/ts-solution-setup");
const _path = require("path");
const _ensuredependencies = require("../../utils/ensure-dependencies");
const _generatorutils = require("../../utils/generator-utils");
const _versions = require("../../utils/versions");
const _init = require("../init/init");
const _detectuiframework = require("../../utils/detect-ui-framework");
function vitestGenerator(tree, schema, hasPlugin = false) {
    return vitestGeneratorInternal(tree, _extends._({
        addPlugin: false
    }, schema), hasPlugin);
}
async function vitestGeneratorInternal(tree, schema, hasPlugin = false) {
    var // Setting default to jsdom since it is the most common use case (React, Web).
    // The @nx/js:lib generator specifically sets this to node to be more generic.
    _schema;
    var _testEnvironment;
    (_testEnvironment = (_schema = schema).testEnvironment) != null ? _testEnvironment : _schema.testEnvironment = 'jsdom';
    const tasks = [];
    const { root, projectType: _projectType } = (0, _devkit.readProjectConfiguration)(tree, schema.project);
    var _schema_projectType;
    const projectType = (_schema_projectType = schema.projectType) != null ? _schema_projectType : _projectType;
    var _schema_uiFramework;
    const uiFramework = (_schema_uiFramework = schema.uiFramework) != null ? _schema_uiFramework : await (0, _detectuiframework.detectUiFramework)(schema.project);
    const isRootProject = root === '.';
    tasks.push(await (0, _js.initGenerator)(tree, _extends._({}, schema, {
        skipFormat: true
    })));
    const initTask = await (0, _init.default)(tree, {
        skipFormat: true,
        addPlugin: schema.addPlugin
    });
    tasks.push(initTask);
    tasks.push((0, _ensuredependencies.ensureDependencies)(tree, _extends._({}, schema, {
        uiFramework
    })));
    (0, _generatorutils.addOrChangeTestTarget)(tree, schema, hasPlugin);
    if (!schema.skipViteConfig) {
        if (uiFramework === 'angular') {
            const relativeTestSetupPath = (0, _devkit.joinPathFragments)('src', 'test-setup.ts');
            const setupFile = (0, _devkit.joinPathFragments)(root, relativeTestSetupPath);
            if (!tree.exists(setupFile)) {
                tree.write(setupFile, `import '@analogjs/vitest-angular/setup-zone';

import {
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting,
} from '@angular/platform-browser-dynamic/testing';
import { getTestBed } from '@angular/core/testing';

getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting()
);
`);
            }
            (0, _generatorutils.createOrEditViteConfig)(tree, {
                project: schema.project,
                includeLib: false,
                includeVitest: true,
                inSourceTests: false,
                imports: [
                    `import angular from '@analogjs/vite-plugin-angular'`
                ],
                plugins: [
                    'angular()'
                ],
                setupFile: relativeTestSetupPath,
                useEsmExtension: true
            }, true);
        } else if (uiFramework === 'react') {
            (0, _generatorutils.createOrEditViteConfig)(tree, {
                project: schema.project,
                includeLib: projectType === 'library',
                includeVitest: true,
                inSourceTests: schema.inSourceTests,
                rollupOptionsExternal: [
                    "'react'",
                    "'react-dom'",
                    "'react/jsx-runtime'"
                ],
                imports: [
                    schema.compiler === 'swc' ? `import react from '@vitejs/plugin-react-swc'` : `import react from '@vitejs/plugin-react'`
                ],
                plugins: [
                    'react()'
                ],
                coverageProvider: schema.coverageProvider
            }, true);
        } else {
            (0, _generatorutils.createOrEditViteConfig)(tree, _extends._({}, schema, {
                includeVitest: true,
                includeLib: projectType === 'library'
            }), true);
        }
    }
    createFiles(tree, schema, root);
    updateTsConfig(tree, schema, root, projectType);
    const coverageProviderDependency = getCoverageProviderDependency(schema.coverageProvider);
    const installCoverageProviderTask = (0, _devkit.addDependenciesToPackageJson)(tree, {}, coverageProviderDependency);
    tasks.push(installCoverageProviderTask);
    // Setup workspace config file (https://vitest.dev/guide/workspace.html)
    if (!isRootProject && !tree.exists(`vitest.workspace.ts`) && !tree.exists(`vitest.workspace.js`) && !tree.exists(`vitest.workspace.json`) && !tree.exists(`vitest.projects.ts`) && !tree.exists(`vitest.projects.js`) && !tree.exists(`vitest.projects.json`)) {
        tree.write('vitest.workspace.ts', `export default ['**/*/vite.config.{ts,mts}', '**/*/vitest.config.{ts,mts}'];`);
    }
    if (!schema.skipFormat) {
        await (0, _devkit.formatFiles)(tree);
    }
    return (0, _devkit.runTasksInSerial)(...tasks);
}
function updateTsConfig(tree, options, projectRoot, projectType) {
    const setupFile = tryFindSetupFile(tree, projectRoot);
    if (tree.exists((0, _devkit.joinPathFragments)(projectRoot, 'tsconfig.spec.json'))) {
        (0, _devkit.updateJson)(tree, (0, _devkit.joinPathFragments)(projectRoot, 'tsconfig.spec.json'), (json)=>{
            var _json_compilerOptions_types, _json_compilerOptions;
            if (!((_json_compilerOptions = json.compilerOptions) == null ? void 0 : (_json_compilerOptions_types = _json_compilerOptions.types) == null ? void 0 : _json_compilerOptions_types.includes('vitest'))) {
                var _json_compilerOptions1;
                if ((_json_compilerOptions1 = json.compilerOptions) == null ? void 0 : _json_compilerOptions1.types) {
                    json.compilerOptions.types.push('vitest');
                } else {
                    var _json;
                    var _compilerOptions;
                    (_compilerOptions = (_json = json).compilerOptions) != null ? _compilerOptions : _json.compilerOptions = {};
                    json.compilerOptions.types = [
                        'vitest'
                    ];
                }
            }
            if (setupFile) {
                var _json_files;
                json.files = [
                    ...(_json_files = json.files) != null ? _json_files : [],
                    setupFile
                ];
            }
            return json;
        });
        (0, _devkit.updateJson)(tree, (0, _devkit.joinPathFragments)(projectRoot, 'tsconfig.json'), (json)=>{
            if (json.references && !json.references.some((r)=>r.path === './tsconfig.spec.json')) {
                json.references.push({
                    path: './tsconfig.spec.json'
                });
            }
            return json;
        });
    } else {
        (0, _devkit.updateJson)(tree, (0, _devkit.joinPathFragments)(projectRoot, 'tsconfig.json'), (json)=>{
            var _json_compilerOptions_types, _json_compilerOptions;
            if (!((_json_compilerOptions = json.compilerOptions) == null ? void 0 : (_json_compilerOptions_types = _json_compilerOptions.types) == null ? void 0 : _json_compilerOptions_types.includes('vitest'))) {
                var _json_compilerOptions1;
                if ((_json_compilerOptions1 = json.compilerOptions) == null ? void 0 : _json_compilerOptions1.types) {
                    json.compilerOptions.types.push('vitest');
                } else {
                    var _json;
                    var _compilerOptions;
                    (_compilerOptions = (_json = json).compilerOptions) != null ? _compilerOptions : _json.compilerOptions = {};
                    json.compilerOptions.types = [
                        'vitest'
                    ];
                }
            }
            return json;
        });
    }
    let runtimeTsconfigPath = (0, _devkit.joinPathFragments)(projectRoot, projectType === 'application' ? 'tsconfig.app.json' : 'tsconfig.lib.json');
    if (options.runtimeTsconfigFileName) {
        runtimeTsconfigPath = (0, _devkit.joinPathFragments)(projectRoot, options.runtimeTsconfigFileName);
        if (!tree.exists(runtimeTsconfigPath)) {
            throw new Error(`Cannot find the specified runtimeTsConfigFileName ("${options.runtimeTsconfigFileName}") at the project root "${projectRoot}".`);
        }
    }
    if (tree.exists(runtimeTsconfigPath)) {
        (0, _devkit.updateJson)(tree, runtimeTsconfigPath, (json)=>{
            if (options.inSourceTests) {
                var _json_compilerOptions;
                var _types;
                ((_types = (_json_compilerOptions = json.compilerOptions).types) != null ? _types : _json_compilerOptions.types = []).push('vitest/importMeta');
            } else {
                const uniqueExclude = new Set([
                    ...json.exclude || [],
                    'vite.config.ts',
                    'vite.config.mts',
                    'vitest.config.ts',
                    'vitest.config.mts',
                    'src/**/*.test.ts',
                    'src/**/*.spec.ts',
                    'src/**/*.test.tsx',
                    'src/**/*.spec.tsx',
                    'src/**/*.test.js',
                    'src/**/*.spec.js',
                    'src/**/*.test.jsx',
                    'src/**/*.spec.jsx'
                ]);
                json.exclude = [
                    ...uniqueExclude
                ];
            }
            if (setupFile) {
                var _json_exclude;
                json.exclude = [
                    ...(_json_exclude = json.exclude) != null ? _json_exclude : [],
                    setupFile
                ];
            }
            return json;
        });
    } else {
        _devkit.logger.warn(`Couldn't find a runtime tsconfig file at ${runtimeTsconfigPath} to exclude the test files from. ` + `If you're using a different filename for your runtime tsconfig, please provide it with the '--runtimeTsconfigFileName' flag.`);
    }
}
function createFiles(tree, options, projectRoot) {
    const isTsSolutionSetup = (0, _tssolutionsetup.isUsingTsSolutionSetup)(tree);
    const rootOffset = (0, _devkit.offsetFromRoot)(projectRoot);
    (0, _devkit.generateFiles)(tree, (0, _path.join)(__dirname, 'files'), projectRoot, _extends._({
        tmpl: ''
    }, options, {
        projectRoot,
        extendedConfig: isTsSolutionSetup ? `${rootOffset}tsconfig.base.json` : './tsconfig.json',
        outDir: isTsSolutionSetup ? `./out-tsc/vitest` : `${rootOffset}dist/out-tsc`
    }));
}
function getCoverageProviderDependency(coverageProvider) {
    switch(coverageProvider){
        case 'v8':
            return {
                '@vitest/coverage-v8': _versions.vitestCoverageV8Version
            };
        case 'istanbul':
            return {
                '@vitest/coverage-istanbul': _versions.vitestCoverageIstanbulVersion
            };
        default:
            return {
                '@vitest/coverage-v8': _versions.vitestCoverageV8Version
            };
    }
}
function tryFindSetupFile(tree, projectRoot) {
    const setupFile = (0, _devkit.joinPathFragments)('src', 'test-setup.ts');
    if (tree.exists((0, _devkit.joinPathFragments)(projectRoot, setupFile))) {
        return setupFile;
    }
}
const _default = vitestGenerator;
